### PR TITLE
perfdata: add experimental generic collector

### LINKED ISSUE #1350: Use PDH native functions as alternative to Registry Calls
Current Progress:

- [x] ad
- [x] adcs
- [x] adfs
- [x] cache
- [x] cpu
- [x] cpu_info
- [x] cs
- [x] container
- [x] diskdrive
- [x] dfsr
- [x] dhcp
- [x] dns
- [x] exchange
- [x] filetime
- [x] fsrmquota
- [x] hyperv
- [x] iis
- [x] license
- [x] logical_disk
- [x] logon
- [x] memory
- [x] mscluster
- [x] msmq
- [x] mssql
- [ ] netframework
- [x] net
- [ ] nps
- [x] os
- [x] physical_disk
- [x] printer
- [x] process
- [x] remote_fx
- [x] scheduled_task
- [x] service
- [x] smb
- [x] smbclient
- [x] smtp
- [x] system
- [x] tcp
- [x] terminal_services
- [x] textfile
- [x] thermalzone
- [x] time
- [x] update
- [x] vmware

---

Reading the documentation about PerfCounter

* https://learn.microsoft.com/en-us/windows/win32/perfctrs/using-the-registry-functions-to-consume-counter-data
* https://learn.microsoft.com/en-us/windows/win32/perfctrs/using-the-pdh-functions-to-consume-counter-data

Microsoft highly recommends to fetch data via PHD function instead native registry. It seems like that PDH function are more performance. 

We have issues like #724 and the exporter does not work well under load and the mentioned Zabbix exporter is using PDH functions, too. Also datadog and telegraf using PDF library.

Ref: 
* https://github.com/zabbix/zabbix/blob/78774f702f8d222235c6e5a8b27db9490386c2e9/src/go/pkg/win32/pdh.go#L112
* https://github.com/DataDog/datadog-agent/blob/cdc12a377528f0604eaf8025c8377d2bf504762f/pkg/util/winutil/pdhutil/pdhhelper.go#L25
* https://github.com/influxdata/telegraf/blob/4c4b821aaa22cc48dc91412c10d2ef8e3fb9c7e3/plugins/inputs/win_perf_counters/pdh.go#L280

We should look into it, since the Registry supports V1 PerfCounter only, while the PDH functions support both V1 and V2.

Since telegraph is using the same OSS license, we should consider to use telegraf libraries as base instead starting from scratch. The [OTEL collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/winperfcounters/internal/third_party/telegraf/win_perf_counters) is doing the same.

image

### LINKED ISSUE #1414: Custom performance counters
Hate to dig up this old one: 

- https://github.com/prometheus-community/windows_exporter/issues/585

But is the advice still to use the text file collector.

I've got a few things I want to do that are available in Perfmon and I can't see a way to extend this for custom metrics easily. This would be such a handy feature not having to have multiple metrics endpoints.

### PR DESCRIPTION
**Pull request by jkroepke**
Ref #1350 
Fixes #1414

This PR setup a generic performance data collectors based on the users configuration.

Instead existing collectors that is using the perflib, the perfdata collector using the Performance Data Helpers function. See #1350

Implement the Performance Data Helpers as separate collector may have multiple oppunities:

It's easier to compare data between Performance Data Helpers and Performance Data from perflib

Any PerfData custom metrics can be implement now, without having the hard requirement of implement collector of each single subsystem.

Example:

```
--collectors.enabled="perfdata" --collector.perfdata.objects='[
  {"object":"Processor Information","instances":["*"],"counters": {"% Processor Time": {}}},
  {"object":"Memory","counters": {"Cache Faults/sec": {"type": "counter"}}}
]'
```

Produces this:

<details>
<summary>Details</summary>

```
# HELP windows_perfdata_memory_cache_faults_sec 
# TYPE windows_perfdata_memory_cache_faults_sec counter
windows_perfdata_memory_cache_faults_sec 2.3861366e+07
# HELP windows_perfdata_processor_information__processor_time 
# TYPE windows_perfdata_processor_information__processor_time gauge
windows_perfdata_processor_information__processor_time{instance="0,0"} 1.7665375e+11
windows_perfdata_processor_information__processor_time{instance="0,1"} 1.797465625e+11
windows_perfdata_processor_information__processor_time{instance="0,10"} 2.315621875e+11
windows_perfdata_processor_information__processor_time{instance="0,11"} 2.3535375e+11
windows_perfdata_processor_information__processor_time{instance="0,12"} 2.37738125e+11
windows_perfdata_processor_information__processor_time{instance="0,13"} 2.3755578125e+11
windows_perfdata_processor_information__processor_time{instance="0,14"} 2.3753984375e+11
windows_perfdata_processor_information__processor_time{instance="0,15"} 2.3741625e+11
windows_perfdata_processor_information__processor_time{instance="0,16"} 2.3767484375e+11
windows_perfdata_processor_information__processor_time{instance="0,17"} 2.3750421875e+11
windows_perfdata_processor_information__processor_time{instance="0,18"} 2.3664765625e+11
windows_perfdata_processor_information__processor_time{instance="0,19"} 2.370809375e+11
windows_perfdata_processor_information__processor_time{instance="0,2"} 1.997065625e+11
windows_perfdata_processor_information__processor_time{instance="0,20"} 2.373096875e+11
windows_perfdata_processor_information__processor_time{instance="0,21"} 2.3738828125e+11
windows_perfdata_processor_information__processor_time{instance="0,22"} 2.3773328125e+11
windows_perfdata_processor_information__processor_time{instance="0,23"} 2.37351875e+11
windows_perfdata_processor_information__processor_time{instance="0,3"} 1.9896265625e+11
windows_perfdata_processor_information__processor_time{instance="0,4"} 2.2467421875e+11
windows_perfdata_processor_information__processor_time{instance="0,5"} 2.3189421875e+11
windows_perfdata_processor_information__processor_time{instance="0,6"} 2.301384375e+11
windows_perfdata_processor_information__processor_time{instance="0,7"} 2.3463625e+11
windows_perfdata_processor_information__processor_time{instance="0,8"} 2.3394984375e+11
windows_perfdata_processor_information__processor_time{instance="0,9"} 2.360415625e+11
```

</details>

In case it works well, some collectors will be migrate from perflib to PDH, too.

### CODE REVIEW

**Comment by DerfOh** [comment]
> Any PerfData custom metrics can be implement now, without having the hard requirement of implement collector of each single subsystem.

Seems this might resolve #1414 as well

File: Makefile [hunk 1]
```diff
@@ -1,6 +1,6 @@
 export GOOS=windows
 export DOCKER_IMAGE_NAME ?= windows-exporter
-export DOCKER_REPO ?= ghcr.io/prometheus-community
+export DOCKER_REPO ?= docker.io/prometheuscommunity ghcr.io/prometheus-community ghcr.io/prometheus-community
```
Comment (breed808) [inline Makefile:3]: Is this related to the PDH change? It's under one of the "fix lint" commits?

**Review by breed808 — COMMENTED** [review]
Excellent work on getting this implemented! I've left a few minor questions, let me know if they require any clarification.

File: Makefile (hunk 1 above)
Comment (jkroepke) [reply Makefile:3]: no, wrong PR

File: pkg/collector/perfdata/perfdata.go [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+// to the provided prometheus Metric channel.
+func (c *collector) Collect(ctx *types.ScrapeContext, ch chan<- prometheus.Metric) error {
+	if desc, err := c.collect(ctx, ch); err != nil {
+		_ = level.Error(c.logger).Log("failed collecting perfdata metrics", "desc", desc, "err", err)
+		return err
+	}
+	return nil
+}
+
+func (c *collector) collect(_ *types.ScrapeContext, ch chan<- prometheus.Metric) (*prometheus.Desc, error) {
```
[the remark this answers is not in the document]
Comment (jkroepke) [reply pkg/collector/perfdata/perfdata.go]: good point

File: pkg/perflib/const.go [hunk 3]
```diff
@@ -1,10 +1,94 @@
...
+	return supportedCounterTypes[counterType] == prometheus.CounterValue
+}
+
+func IsBaseValue(counterType uint32) bool {
+	return counterType == PERF_SAMPLE_BASE || counterType == PERF_RAW_BASE || counterType == PERF_LARGE_RAW_BASE
+}
+
+func IsElapsedTime(counterType uint32) bool {
+	return counterType == PERF_ELAPSED_TIME
+}
```
[the remark this answers is not in the document]
Comment (jkroepke) [reply pkg/perflib/const.go:86]: I'm just copy the original file back from the perfdata project, but we do not have a use for them now. If we have, we can re-add them.

File: pkg/pdh/win_perf_counters.go [hunk 4]
```diff
@@ -0,0 +1,620 @@
...
+		return err
+	}
+	return err
+}
+
+func (m *WinPerfCounters) GetInfo() (map[string]CounterInfos, error) {
+	var wg sync.WaitGroup
+	info := map[string]CounterInfos{}
+
+	// iterate over computers
```
[the remark this answers is not in the document]
Comment (jkroepke) [reply pkg/pdh/win_perf_counters.go:401]: Technically, it's possible to fetch perfdata from remote machines.

Not sure, if we ever want to support this use-case. However I would avoid to change code here. Otherwise it's every hard to copy and backport fixes from upstream.

[the remark this answers is not in the document]
Comment (breed808) [reply pkg/pdh/win_perf_counters.go:401]: Sure, keeping the code unchanged is a good reason. Perhaps add a comment with the context you've provided somewhere in or above that function?

Just seeking to avoid confusion in the future whenever the code needs to be read.

File: pkg/pdh/win_perf_counters.go [hunk 5]
```diff
@@ -0,0 +1,620 @@
...
+		}(hostCounterInfo)
+	}
+
+	wg.Wait()
+	return acc, nil
+}
+
+func (m *WinPerfCounters) gatherComputerCounters(hostCounterInfo *hostCountersInfo) (Accumulator, error) {
+	var value float64
+	var err error
```
[the remark this answers is not in the document]
Comment (jkroepke) [reply pkg/pdh/win_perf_counters.go:504]: I guess, they to avoid re-declaration in each single loop.

File: pkg/pdh/win_perf_counters.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jkroepke) [reply pkg/pdh/win_perf_counters.go:470]: I added a comment on the package level, is that sufficent?

[the remark this answers is not in the document]
Comment (breed808) [reply pkg/pdh/win_perf_counters.go:470]: Yep that's great :+1:
